rust/worker/src/execution/orchestration/get.rs

```diff
@@ -0,0 +1,327 @@
+            GetError::Channel(e) => e.code(),
+            GetError::FetchLog(e) => e.code(),
+            GetError::FetchSegment(e) => e.code(),
+            GetError::Filter(e) => e.code(),
+            GetError::Limit(e) => e.code(),
+            GetError::Panic(_) => ErrorCodes::Aborted,
+            GetError::Projection(e) => e.code(),
+            GetError::Result(_) => ErrorCodes::Internal,
+        }
+    }
+}
+
+impl<E> From<TaskError<E>> for GetError
+where
+    E: Into<GetError>,
+{
+    fn from(value: TaskError<E>) -> Self {
+        match value {
+            TaskError::Panic(e) => GetError::Panic(e.unwrap_or_default()),
+            TaskError::TaskFailed(e) => e.into(),
+        }
+    }
+}
+
+type GetOutput = ProjectionOutput;
+
+type GetResult = Result<GetOutput, GetError>;
+
+#[derive(Debug)]
+pub struct GetOrchestrator {
+    // Orchestrator parameters
+    dispatcher: ComponentHandle<Dispatcher>,
+    queue: usize,
+
+    // Fetch logs and segments
+    fetch_log: FetchLogOperator,
+    fetch_segment: FetchSegmentOperator,
+
+    // Fetch output
+    fetch_log_output: Option<FetchLogOutput>,
+    fetch_segment_output: Option<FetchSegmentOutput>,
+
+    // Pipelined operators
+    filter: FilterOperator,
+    limit: LimitOperator,
+    projection: ProjectionOperator,
+
+    // Result channel
+    result_channel: Option<Sender<GetResult>>,
+}
+
+impl GetOrchestrator {
+    pub fn new(
+        dispatcher: ComponentHandle<Dispatcher>,
+        queue: usize,
+        fetch_log: FetchLogOperator,
+        fetch_segment: FetchSegmentOperator,
+        filter: FilterOperator,
+        limit: LimitOperator,
+        projection: ProjectionOperator,
+    ) -> Self {
+        Self {
+            dispatcher,
+            queue,
+            fetch_log,
+            fetch_segment,
+            fetch_log_output: None,
+            fetch_segment_output: None,
+            filter,
+            limit,
+            projection,
+            result_channel: None,
+        }
+    }
+
+    pub async fn run(mut self, system: System) -> GetResult {
+        let (tx, rx) = oneshot::channel();
+        self.result_channel = Some(tx);
+        let mut handle = system.start_component(self);
+        let result = rx.await;
+        handle.stop();
+        result?
+    }
+
+    // Cleanup the task result and produce the output if any
+    // Terminate the orchestrator if there is any error
+    fn cleanup_response<O, E>(
+        &mut self,
+        ctx: &ComponentContext<Self>,
+        message: TaskResult<O, E>,
+    ) -> Option<O>
+    where
+        E: Into<GetError>,
+    {
+        match message.into_inner() {
+            Ok(output) => Some(output),
+            Err(err) => {
+                self.terminate_with_error(ctx, err.into());
+                None
+            }
+        }
+    }
+
+    fn terminate_with_error(&mut self, ctx: &ComponentContext<Self>, err: GetError) {
+        tracing::error!("Error running orchestrator: {}", err);
```

Comment:
Can we shove the trace into the util method, get rid of the method here and move cleanup_response into a common too?

Reply:
maybe we can trait this and cleanup_response - discussed offline

Reply:
I'll move this to the trait in the future